django/core/cache/backends/memcached.py

```diff
@@ -84,12 +84,9 @@ def delete(self, key, version=None):
         self._cache.delete(key)
 
     def get_many(self, keys, version=None):
-        new_keys = [self.make_key(x, version=version) for x in keys]
-        ret = self._cache.get_multi(new_keys)
-        if ret:
-            m = dict(zip(new_keys, keys))
-            return {m[k]: v for k, v in ret.items()}
-        return ret
+        key_map = {self.make_key(key, version=version): key for key in keys}
+        ret = self._cache.get_multi(key_map.keys())
```

Comment:
You can pass `key_map` directly as `dict.__iter__` yields keys.

Reply:
Unfortunately, this is not possible due to a bug in the `pylibmc` Memcache client.

If you pass a `dict` object directly into `pylibmc`'s `get_multi`, its C implementation will try to get the object's length using [`PySequence_Length()`](https://docs.python.org/3/c-api/sequence.html#c.PySequence_Length) rather than [`PyObject_Length()`](https://docs.python.org/3/c-api/object.html#c.PyObject_Length).

https://github.com/lericson/pylibmc/blob/8e0a86b39af19e4ea64bff650ebdee8752bde6f5/src/_pylibmcmodule.c#L1772

This triggers a `TypeError` in Python, resulting in the following stack trace:

```
Traceback (most recent call last):
  File "[snip]/django/tests/cache/tests.py", line 313, in test_get_many
    self.assertEqual(cache.get_many(['a', 'c', 'd']), {'a': 'a', 'c': 'c', 'd': 'd'})
  File "[snip]/lib/python3.6/site-packages/Django-2.2.dev20180706090656-py3.6.egg/django/core/cache/backends/memcached.py", line 88, in get_many
    ret = self._cache.get_multi(key_map)
TypeError: object of type 'dict' has no len()
```

[This is a known, open issue in CPython](https://bugs.python.org/issue32500), so it seems like there isn't much we can do here. :(

(Note that passing `dict` does work fine if you use the `python-memcached` client.)

Reply:
Makes sense, thanks for the investigation!